src/crewai/tools/base_tool.py

```diff
@@ -0,0 +1,184 @@
+from abc import ABC, abstractmethod
+from typing import Any, Callable, Type, get_args, get_origin
+
+from langchain_core.tools import StructuredTool
+from pydantic import BaseModel, ConfigDict, Field, validator
+from pydantic import BaseModel as PydanticBaseModel
+
+
+class BaseTool(BaseModel, ABC):
+    class _ArgsSchemaPlaceholder(PydanticBaseModel):
+        pass
+
+    model_config = ConfigDict()
+
+    name: str
+    """The unique name of the tool that clearly communicates its purpose."""
+    description: str
+    """Used to tell the model how/when/why to use the tool."""
+    args_schema: Type[PydanticBaseModel] = Field(default_factory=_ArgsSchemaPlaceholder)
+    """The schema for the arguments that the tool accepts."""
+    description_updated: bool = False
+    """Flag to check if the description has been updated."""
+    cache_function: Callable = lambda _args=None, _result=None: True
+    """Function that will be used to determine if the tool should be cached, should return a boolean. If None, the tool will be cached."""
+    result_as_answer: bool = False
+    """Flag to check if the tool should be the final agent answer."""
+
+    @validator("args_schema", always=True, pre=True)
+    def _default_args_schema(
+        cls, v: Type[PydanticBaseModel]
+    ) -> Type[PydanticBaseModel]:
+        if not isinstance(v, cls._ArgsSchemaPlaceholder):
+            return v
+
+        return type(
+            f"{cls.__name__}Schema",
+            (PydanticBaseModel,),
+            {
+                "__annotations__": {
+                    k: v for k, v in cls._run.__annotations__.items() if k != "return"
+                },
+            },
+        )
+
+    def model_post_init(self, __context: Any) -> None:
+        self._generate_description()
+
+        super().model_post_init(__context)
+
+    def run(
+        self,
+        *args: Any,
+        **kwargs: Any,
+    ) -> Any:
+        print(f"Using Tool: {self.name}")
+        return self._run(*args, **kwargs)
+
+    @abstractmethod
+    def _run(
+        self,
+        *args: Any,
+        **kwargs: Any,
+    ) -> Any:
+        """Here goes the actual implementation of the tool."""
+
+    def to_langchain(self) -> StructuredTool:
+        self._set_args_schema()
+        return StructuredTool(
+            name=self.name,
+            description=self.description,
+            args_schema=self.args_schema,
+            func=self._run,
+        )
+
+    @classmethod
+    def from_langchain(cls, tool: StructuredTool) -> "BaseTool":
+        if cls == Tool:
+            return cls(
```

Comment:
I was getting some errors with func not existing on `cls`.

Let's change this to something like this unless you have another suggestion:

```
 @classmethod
  def from_langchain(cls, tool: StructuredTool) -> "BaseTool":
      if cls == Tool:
          if tool.func is None:
              raise ValueError("StructuredTool must have a callable 'func'")
          return Tool(
              name=tool.name,
              description=tool.description,
              args_schema=tool.args_schema,
              func=tool.func,
          )
      raise NotImplementedError(f"from_langchain not implemented for {cls.__name__}")
      ```

Reply:
sweet! just made the change. thanks for reviewing!